### ISSUE TITLE
2015.8.7 pkg.installed problem with version parameter

### ISSUE THREAD

**Issue by alexxannar** [saltstack/salt#31619]
### Description of Issue/Question
Since update to 2015.8.7 version parameter in  pkg.installed is not working correctly. I've tryed to set version by many ways ( 2.7.0, 2.7.0-1, 2:2.7.0-1.el6  etc) and everytime i get this:

The following packages failed to install/update: postfix=2.7.0

Before update all was ok.

### Setup
```sls
install_postfix:
  pkg.installed:
    - name: postfix
    {% if grains['os'] == 'CentOS' %}
    - version: 2.7.0-1
    {% endif %}
```
repos:
```yaml
[root@mymaster postfix]# salt 'myminion' pkg.list_repo_pkgs postfix
myminion:
    ----------
    base:
        ----------
        postfix:
            - 2:2.6.6-6.el6_5
    mycustomrepo:
        ----------
        postfix:
            - 2:2.7.0-1.el6
```

### Steps to Reproduce Issue
Just run sls as is.
log has only this:
```
2016-03-02 16:06:05,329 [salt.state                                                ][ERROR   ][30534] The following packages failed to install/update: postfix=2.7.0-1
```
### Versions Report
```yaml
Salt Version:
           Salt: 2015.8.7

Dependency Versions:
         Jinja2: unknown
       M2Crypto: Not Installed
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.5.0
         Python: 2.6.6 (r266:84292, Jul 23 2015, 15:22:56)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: Not Installed
       cherrypy: 3.2.2
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: 1.4.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: 0.3.7
          smmap: Not Installed
        timelib: Not Installed

System Versions:
           dist: centos 6.7 Final
        machine: x86_64
        release: 2.6.32-573.el6.x86_64
         system: CentOS 6.7 Final
```

**Comment by jfindlay** [comment]
@alexxannar, thanks for reporting.  What version of salt did you upgrade from?  Similar to #30792.

**Comment by ssgward** [comment]
@alexxannar - I can't even install that version using yum. Is there a specific repo that needs to be added to get that version of postfix?

**Comment by seanjnkns** [comment]
I have this exact same issue after updating from 2015.8.1 on CentOS 7.2 to 2015.8.7.  Having something like:

bind-utils:
  pkg.installed:
    - version: 32:9.9.4-29.el7_2.2

Fails.  But if I yum install it and then re-run the check, it's happy.  Somewhere between 2015.8.1 to 2015.8.7 someone added the epoch into the version verification, but that doesn't work for installation.

I'll try and track down the commit that broke this and see if I can do a PR or preferably suggest a revert of that commit til it's properly fixed.  In either case, this issue is forcing me back to 2015.8.1 on our production machines again as we also need the pillar_merge_list fix I created a while ago.

**Comment by alexxannar** [comment]
@jfindlay upgraded from 2015.8.5
@ssgward yep as i described in my example, this is custom repository

**Comment by seanjnkns** [comment]
Here are the results of the git bisect that shows the bad commit:

e1211ed89f621e32b4c8fe006dddfcbe73b8c646 is the first bad commit
commit e1211ed89f621e32b4c8fe006dddfcbe73b8c646
Author: Daniel Wallace <danielwallace@gtmanfred.com>
Date:   Mon Feb 8 13:22:41 2016 -0600

    include possible epoch in version for rpm
    
    Fixes #31014

:040000 040000 2bc5819fee3c73d4f42e7efdeffa1683d0051289 caf95fc7a017f0f7bcd0aeb306eb2b0b8b1e6242 M	salt

**Comment by seanjnkns** [comment]
This would work for dnf and possibly other systems that allow you to specify the epoch in the version name for installation, but not for yum (ie, RHEL/CentOS and others).

@jfindlay, I'll submit a PR to have this reverted as this commit needs a lot more added to it for distro checks, etc... to make this effective.  Also, even though dnf is available for installation on RHEL/CentOS 7.x, salt appears to still utilize yum.

**Comment by seanjnkns** [comment]
Correction to my last statement.  You can specify the epoch on the version installation, but it looks like it's not being properly included for installation, just verification after the fact.  In either case, I'll submit a PR request to have this reverted or see if I can find the other piece(s) that need the epoch included for installation too.

**Comment by seanjnkns** [comment]
Here's an example from a salt-call state.highstate -l debug and the command salt tried to run that failed:

[ERROR   ] Command 'yum -y     install "ksm-10:2.4.0-1.el7.centos"' failed with return code: 1

And, if I run that myself, it fails too.  But, if I change it to be:

yum -y     install "ksm-10:2.4.0-1.el7.centos.x86_64", then it works.  However, that still doesn't work if I change the version in salt itself to contain the architecture.  It'll then install, but fail verification.

**Comment by jfindlay** [comment]
@seanjnkns, if you're going to be working in this area, you should check in with @gtmanfred and/or @terminalmage.

**Comment by terminalmage** [comment]
Fixed in https://github.com/saltstack/salt/pull/31660.

**Comment by terminalmage** [comment]
Actually, my fix just removes the epoch, I'll look at adding the arch instead.

**Comment by seanjnkns** [comment]
Great.  I got side-tracked for a bit, otherwise was going to submit a PR to remove the epoch as a temporary solution.

**Comment by terminalmage** [comment]
@seanjnkns I decided to go with my fix as-is and not try to add the arch. The reason for this is that in Salt we don't make a distinction between a ``noarch`` package and a package that matches the arch of the OS. So, we won't know for certain which arch to append to the end of the package string. Leaving off the epoch when performing the ``yum install`` will still work, the only corner case that could trip up this logic would be if there were multiple versions of the same package available, which were identical with the exception of the epoch, and you were trying to install the older of the two.

**Comment by seanjnkns** [comment]
@terminalmage you doing PRs on this for 2015.5, 2015.8 and the development branch?  Cause I may just patch out the epoch on my production machines to avoid creating a ton of work for myself to role everything back to 2015.8.5 or before.

**Comment by terminalmage** [comment]
No, just the one pull request against 2015.5. We use a merge-forward workflow, so this fix will be merged from 2015.5 into 2015.8, then from 2015.8 into 2016.3, and then 2016.3 into develop.

We do it this way so that we only need a single PR. The pull request is opened against the oldest branch where the bug exists.
